src/sqlfluff/rules/L039.py

```diff
@@ -119,4 +124,24 @@ def _eval(self, context: RuleContext) -> Optional[List[LintResult]]:
                     )
                 )
 
-        return violations or None
+        final_violations = []
+        if violations:
+            # Check each violation. If any of its fixes uses the same anchor
+            # as a previously returned fix, discard it. The linter can't handle
+            # applying fixes like this. Skipping this issue is okay because it
+            # will be detected and fixed during the next linter pass.
```

Comment:
Why is this code in L039? Feels like core code that should be in BaseRule.

Reply:
We don't want `BaseRule` to discard when multiple fixes have the same anchor, because we consider it an error (the `is_valid` check).

L039 had duplicate anchors on this new test case (extracted from one of the `.sql` fixtures):
```
test_excess_space_cast:
  fail_str: |
    select
        '1'    ::   INT as id1,
        '2'::int as id2
    from table_a
  fix_str: |
    select
        '1'::INT as id1,
        '2'::int as id2
    from table_a
```

L039 wants to make two fixes to the line `'1'    ::   INT as id1,`.
1. Replace excessively long whitespace with a single whitespace: "    " -> " " (2 places)
2. Entirely **remove** the whitespace around the `::`.

Thus, it's trying to both **replace** and **delete** the same whitespace. If we return both fixes, the core linter will (appropriately) complain and discard **both** fixes. This bookkeeping ensures that both get fixed, but it's split across two passes through the linter loop. There may be a smarter way to do this, but this approach seems reasonable. I'm trying really hard not to do big rewrites of existing rule code during these PRs -- the goal is to eliminate the critical errors but try and avoid going down a 🐰 hole.

Reply:
Thanks makes sense. Maybe add a comment saying, something like:

This rule works in two steps to remove unnecessary white space:
1. Replace duplicate whitespace to one single whitespace
2. Remove single white spaces if needed.

This can result in two delete being applied to same segment so area so check for that and replace with single delete.